### ISSUE TITLE
Having some questions

### ISSUE THREAD

**Issue by JSteunou** [WebReflection/hyperHTML#14]
Hey, first :clap: very nice work!

I have some questions about this project : 

* I did play with bel + morphdom (morphdom author also says DOM is not slow :) and I agree) the drawback being each call to render actually render the view, (re-)creating a new DOM, then I'm using morphdom to patch the actual DOM. It's faster than some older practice and free the user to manually update the view on state change or writing data binding, but still not optimum as you pointed. But, is your way too much in disfavour of memory usage, by keeping a ref of the statics on the node?
* Would it possible to add some hooks / callback / options to prevent some node to be updated? Use case: using a node as a container for other view, closing / switching / ... I very like this possibility from morphdom.
* Does it play well with svg?

**Comment by WebReflection** [comment]
> is your way too much in disfavour of memory usage, by keeping a ref of the statics on the node?

Not at all, quite the opposite. Nothing is held by statics template references, it's the other way around.
When a node related to a render has no references, the GC is free to get rid of it.
Wires are also weakly related, but not with statics templates, only with instances.

TL;DR this library adds **zero** overhead to the memory needed regardless to render those DOM nodes, since the moment these are used is also the moment these exists regarldess, while statics templates are there forever but you won't create new strings each time, these are always the same, statics indeed so again, nothing added through this library.



> Would it possible to add some hooks / callback / options to prevent some node to be updated?

It's already there. Nodes that don't change don't get updates. You can see in console with your eyes.

Try [this example](https://webreflection.github.io/hyperHTML/test/incremental.html), to name one, and see only the last node is inserted, everything before it doesn't get touched at all.

> Does it play well with svg?

It's called hyper**HTML** for a reason 😋

Truth is, I don't know, and I probably won't add SVG support in here, but I'll eventually consider supporting SVG, together with already supported NodeJS backend, in [viperHTML](https://github.com/WebReflection/viperHTML), which seems more appropriate being universal.

Here I'll keep things as robust and simple as possible.

Hope I've answered your questions.

**Comment by JSteunou** [comment]
I did not expect a response this fast that is nice ! 

About memory I think I understood but maybe my question was not well written.

You are attaching a reference to the static string template on the node using expendo attribute. In a very big application with a lot of view (talking about hundreds here), this could increase memory usage. I understand that this form of caching help a lot regarding the performance, it is always a matter of cpu versus memory. But do you know at what point, which kind or app size, this caching optimization could bloat the memory while indeed keeping the rendering fast ? Comparing to an app that is slower because it would load and render the template on each update.

About the hook/callback on update the trick part is when you are mixing an engine like hyperHTML with manual DOM manipulation. Again the use case is when you are manually mounting/adding/showing (choose the best verb depending the framework you are using) view inside another. You could add a component into an empty node but if the parent view is rendered again and the template string return an empty node it will remove the component instead of keeping it and you would have to manually add it again on each render.

**Comment by WebReflection** [comment]
> In a very big application with a lot of view (talking about hundreds here), this could increase memory usage.

I don't think this is true. I also chose on purpose to avoid WeakMap due GC pressure, taking into accoutn from the scratch gazillion nodes indeed.

Here the thing: templates are immutable and held forever in memory.

If you create a template, that is. It takes the memory up to N bytes and sticks with it.

If you use the same template with 100.000 different functions, you won't have 100.000 * template memory consumption, you'll have template memory consumption, the rest is about your 100.000 objects that will consume memory with or without a template reference.

That reference, is not even a leak issue because like I've said, templates are there to stay forever as unique and statics.

**Comment by WebReflection** [comment]
I give you an example:

```js
((statics, ...changes) => statics)`a${1}c` ===
((statics, ...changes) => statics)`a${2}c`;

// true
```

The `statics` bit is the immutable, static, Array, which refers to the exact same template literals, even if you write it twice in different places.

It doesn't matter what changes in between, those are dynamic bits in between, but the length and the content of the template is the same forever.

TL;DR that script uses the memory it takes to held an immutable/frozen `['a','c']`, and you can reuse that template everywhere you want, no extra memory will be held on RAM, it's just parsing time, further on discarded.

Got it? If not, I'd like to see any real-world concrete example of your leak issue, thanks.

**Comment by WebReflection** [comment]
>  You could add a component into an empty node but if the parent view is rendered again and the template string return an empty node it will remove the component instead of keeping it and you would have to manually add it again on each render.

No idea what you are referring to. Here an example of your specific case/issue would be nice.

All I can say is no, `hyperHTML` allows composing wires and bound nodes without any issue.

```js
const wholeBody = hyperHTML.bind(document.body);

function update() {
  wholeBody`${[
    hyperHTML.wire(header),
    hyperHTML.wire(main),
    hyperHTML.wire(footer)
  ]}`;
}
```

That's it, an entire body served as `hyperHTML`, and it works on NodeJS too foor prime time serving via [viperHTML](https://github.com/WebReflection/viperHTML) using exact same templates and logic.

**Comment by WebReflection** [comment]
an example of multiple wire is in the [hyperHTML TodoMVC app](https://github.com/WebReflection/hypermvc/blob/master/js/app.js).

Wires can be [also inside views](https://github.com/WebReflection/hypermvc/blob/master/js/views/main.js).

I hope you now better understand how hyperHTML works and why it's different from everything else.

Here in case you have more questions.

Regards.

**Comment by JSteunou** [comment]
Thank you for your time that is very clear but again I see I failed to explain my points clearly. I will take the time to ask and explain with examples when I go back next week.

**Comment by JSteunou** [comment]
I'm back :)

So, about my first issue around memory, I got my assertion wrong because I though something like this... (full disclosure I'm a webpack + Backbone.Marionette user ;) )

```
var Mn = require('Marionette');

var MyView = Mn.ItemView.extend({
  template: require('./myview.html'),
  onRender: function() {},
  onAttach: function() {},
});
```

...with `myview.html` being a template string parsed by webpack with a loader kind of like [your bash script](https://github.com/WebReflection/viperHTML#html-in-template-literals-doesnt-get-highlighted) when using `hyperHTML` and not when using simple HTML.

I first though the old fashion way was better on memory but less on cpu because template would be loaded each time I need to render the view and garbage collected after. Which is not the case with `hyperHTML` because a ref is kept on the node. **BUT** of course this assertion is false, because in the case of a build with webpack all assets are here at first load, so no cpu neither network load when loading templates. Also the template is not garbage collected as long as the view is *living* because a ref is kept on the `view.template` attribute so there is no cost at adding the same ref on the node too when using `hyperHTML`.

TLDR; `hyperHTML` is keeping a ref of the template on the node, so as long as the node lives, I am pretty sure the associated view lives too. They both share the same ref to the template, so you are right in my use case, `hyperHTML` do not add more pressure on memory.

**Comment by JSteunou** [comment]
About the second question now, I'm kind of in the same case than https://github.com/WebReflection/hyperHTML/issues/17 so I will re-use its issue if you dont mind

**Comment by WebReflection** [comment]
> ...  the template is not garbage collected as long as ...

this is the bit you seem to misunderstand: a template literal is **never** garbage collected, no matter how many times or where you reference it.

```js
function update(tag) {
  tag`
    even if never used,
    this tempalte will never be
    garbage collected
  `;
}
```

**Comment by jfrazzano** [comment]
I saw the same thing.  I progressively removed updated strings from my dom during update, I then archived the old copies.  I found the same strings in window.HTMLcollectionAll… even days after my archive was off to data base and my object was long since wiped, lived;

**Comment by WebReflection** [comment]
not sure what @jfrazzano is referring to, but what I've said is part of the standard **and** the reason I've chosen that pattern in `hyperHTML`: no matter what I do, template literals are static, immutable, and always available for every realm.

That means if you have 100000000 different template literals, you'll have hard time, with or without `hyperHTML` but since that's not a real-world scenario, everything works just fine, and at speed light.

**Comment by JSteunou** [comment]
Wow I missed that indeed. I was thinking with the old logic where templates are function returning strings.
I should count my views but with template string in memory if the user never reload the webapp it could have a serious footprint

**Comment by WebReflection** [comment]
>  it could have a serious footprint

it doesn't, unless you generate through eval literal templates, which is a whole new world of issues.

Reality is that you'll have a finite amount of template literals that won't fill up the available RAM.

If these do, it means you're targeting the wrong devices/browsers and your architecture needs some polishing regardless `hyperHTML` which biggest merit is not to add RAM consumption to the mix 😉

**Comment by JSteunou** [comment]
Yeah I'm might have over-react a little :p 

But seriously I have a couple hundreds views on a very dynamic app. Views are shown, destroyed as long as the user plays with the app and it's not just going from page 1 to page 2 then go back to page 1. It's a very deep app with tons of features.

The memory question is not directly related to hyperHTML, more to the use of template string (which I already use with bell + morphdom). I dont want a too big memory difference between old fashion way, which allow destroyed views' templates ref to be garbage collected and template string way which of course is still better than virtual or real dom build & patch on each render call. Always a matter of compromise, isn't it? :)
